polynote-frontend/polynote/ui/component/comment.ts

```diff
@@ -0,0 +1,519 @@
+                containerOffset                 // the location of this cell on the page
+                + l.contentWidth                // the width of the content area
+                - l.verticalScrollbarWidth      // don't want to overlay on top of the scrollbar.
+            );
+            const y = containerY + currentY;
+            return [x, y];
+        } else {
+            this.hide(); // should we throw an error instead?
+            return [0, 0];
+        }
+    }
+}
+
+export class CommentButton extends RightGutterOverlay {
+
+    constructor(readonly editor: IStandaloneCodeEditor, readonly range: PosRange, readonly clickCb: () => void) {
+        super(editor, div(['new-comment-button'], []).click((evt) => this.onClick(evt)), range);
+    }
+
+    onClick(evt: Event) {
+        evt.stopPropagation();
+        evt.preventDefault();
+
+        this.clickCb();
+        this.hide();
+    }
+
+    show() {
+        super.show();
+        this.editorListeners.push(this.editor.onDidChangeCursorSelection(() => {
+            const selection = this.editor.getSelection();
+            if (selection) {
+                const model = this.editor.getModel();
+                if (model) {
+                    const range = new PosRange(model.getOffsetAt(selection.getStartPosition()), model.getOffsetAt(selection.getEndPosition()));
+                    if (range.equals(this.range)) {
+                        return;
+                    }
+                }
+            }
+            this.hide(); // if we got here it means the overlay should be hidden.
+        }));
+    }
+}
+
+type CommentContainer = TagElement<"div"> & {comment: CellComment, avatar?: string }
+
+export class CommentUI extends RightGutterOverlay {
+    private readonly commentsEl: TagElement<"div">;
+    private readonly newComment: TagElement<"div">;
+    private newCommentText: TagElement<"textarea">;
+    private readonly commentContainers: CommentContainer[] = [];
+    private highlight: string[] = [];
+
+    constructor(readonly cellId: number, editor: IStandaloneCodeEditor, range: PosRange, readonly currentAuthor: string = "Anonymous", readonly currentAvatar?: string) {
+        super(editor, undefined, range);
+        const [text, controls] = this.commentSubmitter(
+            () => this.doCreate(),
+            () => this.hide()
+        );
+        this.newComment = div(['create-comment', 'comment'], [
+            div(['header'], [
+                span(['avatar'], [currentAvatar]),
+                span(['author'], [currentAuthor]),
+            ]),
+            div(["comment-content"], [
+                this.newCommentText = text,
+                controls
+            ]),
+        ]);
+
+        this.commentsEl = div(['comments-list'], [this.newComment]);
+        this.container.appendChild(this.commentsEl);
+        this.container.classList.add('comment-container');
+
+        this.highlightRange();
+
+        this.editor.onDidChangeModelContent(() => {
```

Comment:
_Note, I refactored the code since writing https://github.com/polynote/polynote/pull/800#discussion_r384783160 so it moved here_

This code is responsible for updating comment ranges based on edits to the cell. I played around with a bunch of ways to do this. I landed on this way, which basically piggybacks on Monaco's range highlights which already do the right thing out of the box. It seemed a bit ugly to rely on on this functionality, but on the other hand it turned out to be the easiest for me to reason about.

100% open to suggestions on how to make it better if you have any ideas for APIs, etc., I might have missed.